map/src/context/TracksRoutingCache.js

```diff
@@ -1,78 +1,201 @@
+                        setStartedRouterJobs((x) => x - 1);
+                        console.debug('updateRouteBeetwenPoints failed', error);
+                    }
+                )
+            );
+
+            if (++started >= MAX_STARTED_ROUTER_JOBS - startedRouterJobs) {
+                break;
+            }
+        }
+    }
+
+    if (started > 0) {
+        return true;
+    }
+
+    ctx.setProcessRouting(false); // all done but a few Promises might still be active
+    return false;
 }
 
-function getRoutingFromCache(track, ctx, map) {
-    for (let i = 0; i < track.points.length - 1; i++) {
-        const start = track.points[i];
-        const end = track.points[i + 1];
-        if (end.geoProfile) {
-            const routingKey = createRoutingKey(start, end, end.geoProfile);
-            const geoCache = ctx.routingCache[routingKey]?.geometry;
-            if (geoCache === STOP_CALC_ROUTING) {
-                let polylineTemp = TrackLayerProvider.createEditableTempLPolyline(start, end, map, ctx);
-                track.layers.addLayer(polylineTemp);
-                track.updateLayers = true;
-                end.geometry = [];
-                updateSelectedRouting(ctx.routingCache[routingKey], polylineTemp, ctx);
-            } else if (geoCache) {
-                end.geometry = geoCache;
+export function effectRefreshTrackWithRouting({ ctx, geoRouter, saveChanges, debouncerTimer }) {
+    let updated = 0;
+    const validKeys = {};
+    const cache = ctx.routingCache;
+    const track = ctx.selectedGpxFile; // ref
+    const points = ctx.selectedGpxFile.points; // ref
+
+    if (points && points.length >= 2) {
+        for (let i = 1; i < points.length; i++) {
+            const startPoint = points[i - 1];
+            const endPoint = points[i];
+            const geoProfile = startPoint.geoProfile;
+            const key = createRoutingKey(startPoint, endPoint, geoProfile);
+
+            if (cache[key]) {
+                const geometry = cache[key].geometry;
+                const tempLine = cache[key].tempLine;
+
+                validKeys[key] = true;
+
+                if (geometry && tempLine) {
+                    // used when tempLine exist (segment recalculation, new points, etc)
+                    refreshTempLine({ ctx, geometry, track, tempLine, color: geoRouter.getColor(startPoint) });
+                    ctx.mutateRoutingCache((o) => o[key] && (o[key].tempLine = null)); // use tempLine only once
+                    endPoint.geometry = geometry; // mutate
+                    updated++;
+                }
             }
         }
     }
-    return track;
+
+    dropOutdatedCache({ ctx, validKeys, killLayers: false });
+
+    if (updated > 0) {
+        requestAnalytics({ ctx, track, debouncerTimer });
+        saveChanges(null, null, null, track); // mutate track with more data and call setSelectedGpxFile({...})
+    }
+
+    return updated > 0;
+}
+
+export function debouncer(f, timerRef, ms) {
```

Comment:
code duplication

Reply:
Not a duplication. export function debouncer was moved from LocalClientTrackLayer to Cache module and are used from both of them.